cpp/src/arrow/type.cc

```diff
@@ -275,6 +256,429 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
   return std::make_shared<Field>(name_, type_, nullable, metadata_);
 }
 
+Field::MergeOptions Field::MergeOptions::Permissive() {
+  MergeOptions options = Defaults();
+  options.promote_nullability = true;
+  options.promote_numeric_width = true;
+  options.promote_integer_float = true;
+  options.promote_integer_decimal = true;
+  options.promote_decimal_float = true;
+  options.promote_date = true;
+  options.promote_time = true;
+  options.promote_duration = true;
+  options.promote_timestamp = true;
+  options.promote_nested = true;
+  options.promote_dictionary = true;
+  options.promote_integer_sign = true;
+  options.promote_large = true;
+  options.promote_binary = true;
+  return options;
```

Comment:
For reading clarity, can we list here the ones that are still left to False in a comment? (eg `promote_dictionary_ordered` is still False, which I think is good)

`promote_decimal` should probably be added as true?

Reply:
Whoops, I didn't update this after adding some of the additional options. I'll document what's not enabled as you suggest, thanks.